File: composeApp/src/commonMain/kotlin/com/marketfinder/core/data/firestore/FirestoreService.kt
package com.marketfinder.core.data.firestore

import kotlinx.coroutines.flow.Flow

/**
 * Firestore Service - Pre-Built Template Component
 *
 * Cross-platform interface for Firebase Firestore.
 * Platform implementations use native Firebase SDKs:
 * - Android: com.google.firebase:firebase-firestore-ktx
 * - iOS: FirebaseFirestore Swift SDK (bridged via factory pattern)
 *
 * Generated repositories should inject this service and delegate
 * Firestore operations to it.
 *
 * Usage:
 * ```
 * class EventRepositoryImpl(
 *     private val firestore: FirestoreService
 * ) : EventRepository {
 *     override suspend fun loadEvents() {
 *         val events = firestore.getCollection("events")
 *         _events.value = events.map { it.toEvent() }
 *     }
 * }
 * ```
 *
 * Setup required:
 * - Android: Add google-services.json to composeApp/
 * - iOS: Add GoogleService-Info.plist to iosApp/iosApp/
 * - iOS: Add Firebase iOS SDK via SPM (FirebaseFirestore package)
 */
interface FirestoreService {
    /** Observe a collection in real-time. Returns a Flow of document maps. */
    fun observeCollection(collectionPath: String): Flow<List<Map<String, Any?>>>

    /** Observe a single document in real-time. */
    fun observeDocument(collectionPath: String, documentId: String): Flow<Map<String, Any?>?>

    /** Get all documents in a collection (one-shot). */
    suspend fun getCollection(collectionPath: String): List<Map<String, Any?>>

    /** Get a single document (one-shot). */
    suspend fun getDocument(collectionPath: String, documentId: String): Map<String, Any?>?

    /** Add a document (auto-generated ID). Returns the document ID. */
    suspend fun addDocument(collectionPath: String, data: Map<String, Any?>): String

    /** Set (create/overwrite) a document at a specific ID. */
    suspend fun setDocument(collectionPath: String, documentId: String, data: Map<String, Any?>)

    /** Update specific fields of a document. */
    suspend fun updateDocument(collectionPath: String, documentId: String, fields: Map<String, Any?>)

    /** Delete a document. */
    suspend fun deleteDocument(collectionPath: String, documentId: String)

    /** Query documents with simple filters. */
    suspend fun queryCollection(
        collectionPath: String,
        field: String,
        op: String,     // "==", "!=", "<", "<=", ">", ">=", "array-contains"
        value: Any
    ): List<Map<String, Any?>>
}

File: iosApp/iosApp/util/firebase/FirestoreService.swift
import Foundation
import FirebaseFirestore
import ComposeApp

class IOSFirestoreService: NSObject, FirestoreService {

    private lazy var db: Firestore = {
        Firestore.firestore()
    }()

    static let shared = IOSFirestoreService()

    private override init() { super.init() }

    // MARK: - Real-time Observe (returns Kotlinx Flow via callback)

    // Note: Real-time observation requires CFlow/callback bridge.
    // For the template, we provide one-shot methods.
    // The code-gen LLM can add real-time wrappers as needed.

    func observeCollection(collectionPath: String) -> any Kotlinx_coroutines_coreFlow{
        // Placeholder — real-time needs CFlow bridge or polling
        fatalError("Use Android for real-time or implement CFlow bridge")
    }

    func observeDocument(collectionPath: String, documentId: String) -> any Kotlinx_coroutines_coreFlow {
        fatalError("Use Android for real-time or implement CFlow bridge")
    }

    func getCollection(collectionPath: String) async throws -> [[String: Any]] {
        let snapshot = try await db.collection(collectionPath).getDocuments()
        return snapshot.documents.map { doc in
            var data = doc.data()
            data["id"] = doc.documentID
            return data
        }
    }

    func getDocument(collectionPath: String, documentId: String) async throws -> [String: Any]? {
        let doc = try await db.collection(collectionPath).document(documentId).getDocument()
        guard var data = doc.data() else { return nil }
        data["id"] = doc.documentID
        return data
    }

    func addDocument(collectionPath: String, data: [String: Any]) async throws -> String {
        let ref = try await db.collection(collectionPath).addDocument(data: data)
        return ref.documentID
    }

    func setDocument(collectionPath: String, documentId: String, data: [String: Any]) async throws {
        try await db.collection(collectionPath).document(documentId).setData(data)
    }

    func updateDocument(collectionPath: String, documentId: String, fields: [String: Any]) async throws {
        try await db.collection(collectionPath).document(documentId).updateData(fields)
    }

    func deleteDocument(collectionPath: String, documentId: String) async throws {
        try await db.collection(collectionPath).document(documentId).delete()
    }

    func queryCollection(collectionPath: String, field: String, op: String, value: Any) async throws -> [[String: Any]] {
        var query: Query = db.collection(collectionPath)
        switch op {
        case "==": query = query.whereField(field, isEqualTo: value)
        case "!=": query = query.whereField(field, isNotEqualTo: value)
        case "<": query = query.whereField(field, isLessThan: value)
        case "<=": query = query.whereField(field, isLessThanOrEqualTo: value)
        case ">": query = query.whereField(field, isGreaterThan: value)
        case ">=": query = query.whereField(field, isGreaterThanOrEqualTo: value)
        case "array-contains": query = query.whereField(field, arrayContains: value)
        default: break
        }
        let snapshot = try await query.getDocuments()
        return snapshot.documents.map { doc in
            var data = doc.data()
            data["id"] = doc.documentID
            return data
        }
    }
}
